project/cyk.py

```diff
@@ -0,0 +1,53 @@
+from pyformlang.cfg import CFG
+
+__all__ = ["cyk"]
+
+
+def cyk(cfg: CFG, word: str) -> bool:
+    """
+    Determines whether the given string can be generated in the given context-free grammar
+
+    Parameters
+    ----------
+    cfg: CFG
+        input grammar
+    word: str
+        word for checking
+
+    Returns
+    -------
+    bool
+        true if word can be generated in the given cfg
+        false otherwise
+
+    """
+    word_len = len(word)
+
+    if not word_len:
+        return cfg.generate_epsilon()
+
+    cnf = cfg.to_normal_form()
+
+    term_productions = [p for p in cnf.productions if len(p.body) == 1]
+    var_productions = [p for p in cnf.productions if len(p.body) == 2]
+
+    dp_matrix = [[set() for _ in range(word_len)]] * word_len
```

Comment:
```python
>>> d = [[set() for _ in range(2)]] * 2
>>> d
[[set(), set()], [set(), set()]]
>>> d[0][0].add(1)
>>> d
[[{1}, set()], [{1}, set()]]
```
Я думаю, вы такого не ожидали.

Лучше сделать так.
```suggestion
    dp_matrix = [[set() for _ in range(word_len)] for _ in range(word_len)]
```

Тогда будет
```python
>>> d = [[set() for _ in range(2)] for _ in range(2)] 
>>> d
[[set(), set()], [set(), set()]]
>>> d[0][0].add(1)
>>> d
[[{1}, set()], [set(), set()]]
```